cmake/options.cmake

```diff
@@ -4,6 +4,54 @@
 # This source code is licensed in accordance with the terms specified in
 # the LICENSE file found in the root directory of this source tree.
 
+# Detect osquery version
+# The OSQUERY_VERSION cache variable will be used if set or not empty
+# Otherwise detect version through git and set it in the OSQUERY_VERSION_AUTODETECTED cache variable. If detection fails, 0.0.0 will be used.
+# Append OSQUERY_VERSION_SUFFIX cache variable to the final version if specified and the version is detected from git.
+# Verify if the final version is composed by three semver components, otherwise fail.
+# Returns the final version in OSQUERY_VERSION_INTERNAL and its components in OSQUERY_VERSION_COMPONENTS
+function(detectOsqueryVersion)
+  set(OSQUERY_VERSION "" CACHE STRING "Overrides osquery version with this value")
+  set(OSQUERY_VERSION_SUFFIX "" CACHE STRING "String to append when the version is automatically detected")
+  set(OSQUERY_VERSION_AUTODETECTED "" CACHE STRING "osquery version autodetected through git. Do not manually set." FORCE)
+  set(osquery_version 0.0.0)
+
+  if(NOT OSQUERY_VERSION)
+    find_package(Git REQUIRED)
+
+    execute_process(
+      COMMAND "${GIT_EXECUTABLE}" describe --tags --always --dirty
+      WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+      OUTPUT_VARIABLE branch_version
+      RESULT_VARIABLE exit_code
+    )
+
+    if(NOT ${exit_code} EQUAL 0)
+      message(WARNING "Failed to detect osquery version. Set it manually through OSQUERY_VERSION or 0.0.0 will be used")
+    else()
+      string(REGEX REPLACE "\n$" "" branch_version "${branch_version}")
+      set(osquery_version ${branch_version})
+      overwrite_cache_variable("OSQUERY_VERSION_AUTODETECTED" "STRING" ${osquery_version})
+
+      if(OSQUERY_VERSION_SUFFIX)
```

Comment:
Well but that is counter intuitive on the description and current behavior of OSQUERY_VERSION.
Currently OSQUERY_VERSION is meant to be THE final version if set, the one that has priority over everything.
As far as I understand your suggestion is because the version suffix, if needed even in the manual case (when setting OSQUERY_VERSION), will be appended by us instead of requiring the user to prepare the version it wants to see.

Reply:
On a second thought though I see where you come from, the meaning of OSQUERY_VERSION really changed in experimental...
Maybe lets first set the record straight with OSQUERY_BUILD_VERSION (which was behaving as OSQUERY_VERSION_SUFFIX in the past), then we have a second look at this.

Reply:
As I think about it, I think it's hard to predict whether people will want `OSQUERY_VERSION` to override `OSQUERY_VERSION_SUFFIX` or vice verse. 

I suspect that this is really deep in the weeds. I bet most builds will use the git version, and some people will want to append a suffix. And beyond that, it mostly doesn't matter.